transport/http2_client.go

```diff
@@ -649,47 +649,63 @@ func (t *http2Client) Close() (err error) {
 	return
 }
 
-func (t *http2Client) GracefulClose() error {
-	t.mu.Lock()
+// goAwayActiveStreams is called when a GoAway frame is received. It assumes a transport lock is held.
+func (t *http2Client) goAwayActiveStreams() {
+	// A client can recieve multiple GoAways from server (look at https://github.com/grpc/grpc-go/issues/1387).
+	// The idea is that the first GoAway will be sent with an ID of MaxInt32 and the second GoAway will be sent after an RTT delay
+	// with the ID of the last stream the server will process.
+	// Therefore, when we get the first GoAway we don't really close any streams. While in case of second GoAway we
+	// close all streams created after the second GoAwayId. This way streams that were in-flight while the GoAway from server
+	// was being sent don't get killed.
+	//
+	// Note: to be backward compatible with servers that will still send only 1 GoAway we'll still try and close streams with ID
+	// greater that GoAwayId sent by the first GoAway.
+
+	n := t.prevGoAwayID
+	if n == 0 && t.nextID > 1 {
+		n = t.nextID - 2
+	}
+	m := t.goAwayID + 2
+	if m == 2 {
+		m = 1
+	}
+	for i := m; i <= n; i += 2 {
+		if s, ok := t.activeStreams[i]; ok {
+			close(s.goAway)
+		}
+	}
+}
```

Comment:
t.prevGoAwayID probably should be set to `end - 2` here.  Otherwise if we get
```
GOAWAY(INT32_MAX)
GOAWAY(0)
```
the second goaway will go through 1B iterations.

OR, we can do something like this, above:

```go
if t.nextID == 1 {  // or len(t.activeStreams) == 0
  return // there are no streams to stop
}

end := t.prevGoAwayID
if end == 0 || end > t.nextID {
  end = t.nextID - 2
}
```